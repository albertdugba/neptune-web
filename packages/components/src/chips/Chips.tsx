import classNames from 'classnames';
import { ReactElement } from 'react';
import { useIntl } from 'react-intl';

import { CommonProps, AriaLabelProperty } from '../common';

import Chip from './Chip';
import messages from './Chips.messages';

export type ChipValue = string | number;

export type Chip = {
  value: ChipValue;
  label: string;
};

export type ChipsProps = CommonProps &
  AriaLabelProperty & {
    /** List of chips with string labels and string/number values */
    chips: Chip[];
    /** Callback which is invoked when a chip is selected or deselected */
    onChange: ({
      isEnabled,
      selectedValue,
    }: {
      isEnabled: boolean;
      selectedValue: ChipValue;
    }) => void;
    /** Used to manage which chips are selected */
    selected: ChipValue | ChipValue[];
    /** Used to activate multi-selection */
    multiple?: boolean;
  };

const Chips = ({
  chips,
  onChange,
  selected,
  'aria-label': ariaLabel,
  className,
  multiple,
}: ChipsProps): ReactElement => {
  const intl = useIntl();

  const isSelected = (value: ChipValue) =>
    Array.isArray(selected) ? selected.includes(value) : selected === value;

  const handleOnChange = (selectedValue: ChipValue, isCurrentlyEnabled: boolean) => {
    onChange({ isEnabled: !isCurrentlyEnabled, selectedValue });
  };

  return (
    <div
      className={classNames('np-chips d-flex', className)}
      aria-label={ariaLabel}
      role={!multiple ? 'radiogroup' : undefined}
    >
      {chips.map((chip) => {
        const chipSelected = isSelected(chip.value);
        return (
          <Chip
            key={chip.value}
            value={chip.value}
            label={chip.label}
            closeButton={{
              'aria-label': intl.formatMessage(messages.ariaLabel, { choice: chip.label }),
            }}
            className={classNames('text-xs-nowrap', {
              'np-chip--selected': chipSelected,
              'p-r-1': multiple && chipSelected,
            })}
            {...(multiple && chipSelected
              ? {
                  onRemove: () => handleOnChange(chip.value, chipSelected),
                  'aria-label': chip.label,
                }
              : {
                  onClick: () => handleOnChange(chip.value, chipSelected),
                  onKeyPress: () => handleOnChange(chip.value, chipSelected),
                  role: !multiple ? 'radio' : undefined,
                  'aria-checked': chipSelected,
                })}
          />
        );
      })}
    </div>
  );
};

export default Chips;
