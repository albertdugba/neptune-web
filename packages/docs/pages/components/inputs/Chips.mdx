import { LiveEditorBlock, GeneratePropsTable } from '../../../utils';
import { Chips } from '@transferwise/components';
import choiceCode from '../../../liveEditorCode/chips.choice.code';
import filterCode from '../../../liveEditorCode/chips.filter.code';

The `Chips` component can be used for making a **choice**, or as a **filter** with multiple options.

## Choice

The user can only select one chip with the `selected` prop as a single value.

<LiveEditorBlock code={choiceCode} scope={{ Chips }} />

---

## Filter

The user can select any number of chips with the `selected` prop as an array.

<LiveEditorBlock code={filterCode} scope={{ Chips }} />
<GeneratePropsTable componentName="Chips" />

---

export const meta = {
  name: 'Chips',
  badge: {
    type: 'new',
    expiryDate: '2021-11-01',
  },
};
