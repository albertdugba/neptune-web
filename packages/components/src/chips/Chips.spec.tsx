import '@testing-library/jest-dom';
import { render, screen, fireEvent } from '../test-utils';

import Chips from '.';

const choiceProps = {
  chips: [
    {
      value: 'accounting',
      label: 'Accounting',
    },
    {
      value: 'payroll',
      label: 'Payroll',
    },
    {
      value: 'reporting',
      label: 'Reporting',
    },
    {
      value: 'payments',
      label: 'Payments',
    },
  ],
  selected: 'payroll',
  onChange: jest.fn(),
  'aria-label': 'Please select a category',
};

const filterProps = {
  chips: [
    {
      value: 'accounting',
      label: 'Accounting',
    },
    {
      value: 'payroll',
      label: 'Payroll',
    },
    {
      value: 'reporting',
      label: 'Reporting',
    },
    {
      value: 'payments',
      label: 'Payments',
    },
  ],
  onChange: jest.fn(),
  'aria-label': 'Filter by categories',
  selected: ['payments'],
  multiple: true,
};

describe('Chips', () => {
  describe('Choice Chips', () => {
    it(`renders as expected`, () => {
      expect(render(<Chips {...choiceProps} />).container).toMatchSnapshot();
    });

    it(`will pass previous state, updated state, and selected value on select`, () => {
      render(<Chips {...choiceProps} />);
      const accounting = screen.getByText('Accounting');
      fireEvent.click(accounting);
      expect(choiceProps.onChange).toHaveBeenCalledWith({
        selectedValue: 'accounting',
        isEnabled: true,
      });
    });
  });

  describe('Filter Chips', () => {
    it(`renders as expected`, () => {
      expect(render(<Chips {...filterProps} />).container).toMatchSnapshot();
    });

    it(`will pass previous state, updated state, and selected value on select`, () => {
      render(<Chips {...filterProps} />);
      const accounting = screen.getByText('Accounting');
      fireEvent.click(accounting);
      expect(filterProps.onChange).toHaveBeenCalledWith({
        selectedValue: 'accounting',
        isEnabled: true,
      });
    });

    it(`will pass previous state, updated state, and cleared value on clear`, () => {
      render(<Chips {...filterProps} />);
      fireEvent.click(screen.getAllByRole('button')[3]);
      expect(filterProps.onChange).toHaveBeenCalledWith({
        selectedValue: 'payments',
        isEnabled: false,
      });
    });
  });
});
