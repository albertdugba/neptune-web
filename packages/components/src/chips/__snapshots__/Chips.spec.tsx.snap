// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Chips Choice Chips renders as expected 1`] = `
<div>
  <div
    aria-label="Please select a category"
    class="np-chips d-flex"
    role="radiogroup"
  >
    <div
      aria-checked="false"
      class="np-chip d-flex align-items-center justify-content-between text-xs-nowrap"
      role="radio"
      tabindex="0"
    >
      <span
        aria-hidden="false"
        class="np-chip-label"
      >
        Accounting
      </span>
    </div>
    <div
      aria-checked="true"
      class="np-chip d-flex align-items-center justify-content-between text-xs-nowrap np-chip--selected"
      role="radio"
      tabindex="0"
    >
      <span
        aria-hidden="false"
        class="np-chip-label"
      >
        Payroll
      </span>
    </div>
    <div
      aria-checked="false"
      class="np-chip d-flex align-items-center justify-content-between text-xs-nowrap"
      role="radio"
      tabindex="0"
    >
      <span
        aria-hidden="false"
        class="np-chip-label"
      >
        Reporting
      </span>
    </div>
    <div
      aria-checked="false"
      class="np-chip d-flex align-items-center justify-content-between text-xs-nowrap"
      role="radio"
      tabindex="0"
    >
      <span
        aria-hidden="false"
        class="np-chip-label"
      >
        Payments
      </span>
    </div>
  </div>
</div>
`;

exports[`Chips Filter Chips renders as expected 1`] = `
<div>
  <div
    aria-label="Filter by categories"
    class="np-chips d-flex"
  >
    <div
      aria-checked="false"
      class="np-chip d-flex align-items-center justify-content-between text-xs-nowrap"
      role="button"
      tabindex="0"
    >
      <span
        aria-hidden="false"
        class="np-chip-label"
      >
        Accounting
      </span>
    </div>
    <div
      aria-checked="false"
      class="np-chip d-flex align-items-center justify-content-between text-xs-nowrap"
      role="button"
      tabindex="0"
    >
      <span
        aria-hidden="false"
        class="np-chip-label"
      >
        Payroll
      </span>
    </div>
    <div
      aria-checked="false"
      class="np-chip d-flex align-items-center justify-content-between text-xs-nowrap"
      role="button"
      tabindex="0"
    >
      <span
        aria-hidden="false"
        class="np-chip-label"
      >
        Reporting
      </span>
    </div>
    <div
      aria-label="Payments"
      class="np-chip d-flex align-items-center justify-content-between text-xs-nowrap np-chip--selected p-r-1"
      tabindex="-1"
    >
      <span
        aria-hidden="true"
        class="np-chip-label"
      >
        Payments
      </span>
      <button
        aria-label="Clear Payments"
        class="np-close-button close btn-link text-no-decoration btn-unstyled m-l-1"
        type="button"
      >
        <span
          aria-hidden="true"
          class="tw-icon tw-icon-cross-circle "
          role="presentation"
        >
          <svg
            fill="currentColor"
            focusable="false"
            height="16"
            width="16"
          >
            <path
              clip-rule="evenodd"
              d="M8 16A8 8 0 108 0a8 8 0 000 16zm2.434-4.434L8 9.13l-2.434 2.435-1.132-1.132L6.87 8 4.434 5.566l1.132-1.132L8 6.87l2.434-2.435 1.132 1.132L9.13 8l2.435 2.434-1.132 1.132z"
              fill-rule="evenodd"
            />
          </svg>
        </span>
      </button>
    </div>
  </div>
</div>
`;
